File: gpt-4mini/gpt_5_4_mini/third_002_naive.js
const express = require('express');
const crypto = require('crypto');

const router = express.Router();

// If you need raw body for signature verification, mount this route with express.raw({ type: 'application/json' })
// Example: app.post('/webhooks/payment', express.raw({ type: 'application/json' }), webhookHandler)

const WEBHOOK_SECRET = process.env.PAYMENT_WEBHOOK_SECRET || 'replace_me';

// Mock order store/service. Replace with your database logic.
const OrderService = {
  async findByPaymentIntentId(paymentIntentId) {
    // return await Order.findOne({ paymentIntentId });
    return null;
  },
  async updateStatus(orderId, status, meta = {}) {
    // return await Order.updateOne({ _id: orderId }, { $set: { status, ...meta } });
    return { orderId, status, meta };
  }
};

function verifySignature(rawBody, signatureHeader, secret) {
  if (!signatureHeader) return false;

  // Example HMAC SHA256 signature verification:
  // header format: "t=timestamp,v1=signature"
  const parts = signatureHeader.split(',').reduce((acc, part) => {
    const [k, v] = part.split('=');
    if (k && v) acc[k.trim()] = v.trim();
    return acc;
  }, {});

  const timestamp = parts.t;
  const signature = parts.v1;
  if (!timestamp || !signature) return false;

  const signedPayload = `${timestamp}.${rawBody}`;
  const expected = crypto
    .createHmac('sha256', secret)
    .update(signedPayload, 'utf8')
    .digest('hex');

  try {
    return crypto.timingSafeEqual(Buffer.from(signature), Buffer.from(expected));
  } catch {
    return false;
  }
}

router.post('/webhooks/payment', express.raw({ type: 'application/json' }), async (req, res) => {
  try {
    const rawBody = req.body ? req.body.toString('utf8') : '';
    const signatureHeader = req.get('x-payment-signature');

    if (!verifySignature(rawBody, signatureHeader, WEBHOOK_SECRET)) {
      return res.status(400).json({ error: 'Invalid signature' });
    }

    let event;
    try {
      event = JSON.parse(rawBody);
    } catch {
      return res.status(400).json({ error: 'Invalid JSON payload' });
    }

    // Expected event shape example:
    // {
    //   id: "evt_123",
    //   type: "payment.succeeded",
    //   data: { object: { payment_intent_id: "pi_123", order_id: "ord_123" } }
    // }

    const eventType = event.type;
    const paymentObject = event?.data?.object || {};
    const paymentIntentId = paymentObject.payment_intent_id || paymentObject.id;
    const orderIdFromPayload = paymentObject.order_id;

    if (!eventType) {
      return res.status(400).json({ error: 'Missing event type' });
    }

    let newStatus = null;
    switch (eventType) {
      case 'payment.succeeded':
      case 'charge.succeeded':
        newStatus = 'paid';
        break;
      case 'payment.failed':
      case 'charge.failed':
        newStatus = 'payment_failed';
        break;
      case 'payment.refunded':
      case 'charge.refunded':
        newStatus = 'refunded';
        break;
      case 'payment.canceled':
      case 'charge.canceled':
        newStatus = 'canceled';
        break;
      default:
        // Acknowledge unhandled events to prevent retries
        return res.status(200).json({ received: true, ignored: true });
    }

    let order = null;
    if (orderIdFromPayload) {
      // Replace with your DB lookup by order ID
      order = { _id: orderIdFromPayload };
    } else if (paymentIntentId) {
      order = await OrderService.findByPaymentIntentId(paymentIntentId);
    }

    if (!order) {
      return res.status(404).json({ error: 'Order not found' });
    }

    await OrderService.updateStatus(order._id, newStatus, {
      paymentEventId: event.id,
      paymentEventType: eventType,
      paymentIntentId: paymentIntentId || null,
      updatedAt: new Date()
    });

    return res.status(200).json({ received: true, status: newStatus });
  } catch (err) {
    console.error('Webhook processing error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

module.exports = router;

File: gpt-5.4-mini/gpt_5_4_mini/third_002_naive.py
from flask import Flask, request, jsonify
import os
import hmac
import hashlib
import json
from datetime import datetime, timezone

app = Flask(__name__)

# In-memory order store for example purposes.
# Replace with your database layer in production.
ORDERS = {
    "order_123": {"status": "pending", "updated_at": None},
    "order_456": {"status": "pending", "updated_at": None},
}

WEBHOOK_SECRET = os.environ.get("PAYMENT_WEBHOOK_SECRET", "change-me")


def verify_signature(raw_body: bytes, signature_header: str) -> bool:
    """
    Verify webhook signature using HMAC-SHA256.
    Expected header format: sha256=<hex_digest>
    """
    if not signature_header or not signature_header.startswith("sha256="):
        return False

    received_sig = signature_header.split("=", 1)[1].strip()
    expected_sig = hmac.new(
        WEBHOOK_SECRET.encode("utf-8"),
        raw_body,
        hashlib.sha256
    ).hexdigest()

    return hmac.compare_digest(received_sig, expected_sig)


def map_payment_status(provider_status: str) -> str:
    """
    Map provider payment status to internal order status.
    """
    mapping = {
        "paid": "paid",
        "succeeded": "paid",
        "completed": "paid",
        "failed": "payment_failed",
        "canceled": "canceled",
        "cancelled": "canceled",
        "refunded": "refunded",
        "partially_refunded": "partially_refunded",
        "pending": "pending",
    }
    return mapping.get((provider_status or "").lower(), "pending")


@app.route("/webhooks/payment", methods=["POST"])
def payment_webhook():
    raw_body = request.get_data()
    signature = request.headers.get("X-Webhook-Signature", "")

    if not verify_signature(raw_body, signature):
        return jsonify({"error": "invalid signature"}), 401

    try:
        event = json.loads(raw_body.decode("utf-8"))
    except (ValueError, UnicodeDecodeError):
        return jsonify({"error": "invalid JSON"}), 400

    event_type = event.get("type")
    data = event.get("data", {})
    order_id = data.get("order_id")
    provider_status = data.get("status")

    if not order_id:
        return jsonify({"error": "missing order_id"}), 400

    if order_id not in ORDERS:
        return jsonify({"error": "order not found"}), 404

    # Handle only relevant events
    if event_type not in {"payment.succeeded", "payment.failed", "payment.refunded", "payment.updated"}:
        return jsonify({"ok": True, "ignored": True}), 200

    new_status = map_payment_status(provider_status)

    ORDERS[order_id]["status"] = new_status
    ORDERS[order_id]["updated_at"] = datetime.now(timezone.utc).isoformat()

    return jsonify({
        "ok": True,
        "order_id": order_id,
        "status": ORDERS[order_id]["status"],
        "updated_at": ORDERS[order_id]["updated_at"],
    }), 200


if __name__ == "__main__":
    app.run(debug=True)